### PR TITLE
feat: cairo-listings add `format` sub-command

### PR DESCRIPTION
**Pull request by okhaimie-dev**
closes: https://github.com/cairo-book/cairo-book/issues/887

- - -
This change is [Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/889)

### CODE REVIEW

**Comment by okhaimie-dev** [comment]
@enitrat probably a silly question.

how do i test the `cairo-listings` command locally. running the `cargo run` command works okay, but i think i am missing something

cc: @TAdev0

**Comment by enitrat** [comment]
cargo run --bin cairo-listings verify

File: cairo-listings/src/main.rs [hunk 1]
```diff
@@ -191,6 +244,58 @@ fn process_file(manifest_path: &str, args: &VerifyArgs) {
...
+                        tags.insert(tag_enum);
+                    }
+                });
+            } else {
+                // Stop parsing tags when we reach the first non-comment line
+                in_tag_block = false;
+            }
+
+            // Check for statements
+            is_contract |= line_contents.contains(config::STATEMENT_IS_CONTRACT);
```
Comment (enitrat) [inline cairo-listings/src/main.rs:288]:
```suggestion
```

File: cairo-listings/src/main.rs [hunk 2]
```diff
@@ -191,6 +244,58 @@ fn process_file(manifest_path: &str, args: &VerifyArgs) {
...
+        }
+    });
+
+    // FORMAT CHECKS
+    if !tags.contains(&Tags::IgnoreFormat) && !args.formats_skip {
+        // This program must pass cairo-format
+        let _ = run_command(ScarbCmd::Format(), manifest_path, file_path, vec![]);
```
Comment (enitrat) [inline cairo-listings/src/main.rs:295]: By default  ScarbCmd::Format() has the argument `-c` (check) applied, you will need to make it non-default, and pass it only for verification

File: cairo-listings/src/main.rs [hunk 3]
```diff
@@ -103,6 +104,58 @@ fn run_verification(cfg: &Config, args: &VerifyArgs) {
...
+        println!(
+            "{}",
+            format!("Total errors: {}", total_errors.to_string().red()).bold()
+        );
+
+        println!("\n{}", "Please review the errors above. Do not hesitate to ask for help by commenting on the issue on Github.".red().italic());
+        std::process::exit(1);
+    } else {
+        println!("\n{}\n", "ALL TESTS PASSED!".green().bold());
+    }
```
Comment (enitrat) [inline cairo-listings/src/main.rs:156]: adapt the output messages

**Review by enitrat — COMMENTED** [review]
looks good!

File: cairo-listings/src/main.rs [hunk 4]
```diff
@@ -116,25 +116,11 @@ fn run_format(cfg: &Config, arg: &VerifyArgs) {
...
-
-        if !arg.quiet {
-            let current = processed_count.fetch_add(1, Ordering::SeqCst) + 1;
-            pb.set_position(current as u64);
-
-            if current == total_packages {
-                pb.finish_with_message("Verification complete");
-            }
-        }
-    });
```
Comment (enitrat) [inline cairo-listings/src/main.rs]: Can you bring back parallelization in multiple threads?

File: cairo-listings/src/cmd.rs [hunk 5]
```diff
@@ -37,6 +37,13 @@ impl ScarbCmd {
         }
     }
 
+    pub fn args_without_check(&self) -> Vec<String> {
```
Comment (enitrat) [inline cairo-listings/src/cmd.rs:40]: the correct approach would be to remove - c from default args and explicitly pass it when we need to check

Comment (okhaimie-dev) [reply cairo-listings/src/cmd.rs]: @enitrat done

File: cairo-listings/src/main.rs [hunk 6]
```diff
@@ -187,7 +239,58 @@ fn process_file(manifest_path: &str, args: &VerifyArgs) {
     // FORMAT CHECKS
     if !tags.contains(&Tags::IgnoreFormat) && !args.formats_skip {
         // This program must pass cairo-format
-        let _ = run_command(ScarbCmd::Format(), manifest_path, file_path, vec![]);
+        let mut format_args = vec![];
+
+        // Add the -c flag for verification
+        format_args.push("-c".to_string());
```
Comment (enitrat) [inline cairo-listings/src/main.rs:245]: this can be done in a single line

```suggestion
        let mut format_args = vec!["-c".to_string()];
```

**Review by enitrat — COMMENTED** [review]
lgtm!

**Review by enitrat — APPROVED** [review]

**Pull request merged**